File: Agents/Interaction/Templates/Debug.cs
namespace Technologai.Templates
{
    public class Debug : Template
    {
        private Agent _agent;
        public Debug(Agent agent)
        {
            Id = "debug";
            Description = "Debug";
            _agent = agent;
        }

        public override Task<bool> Assess(Information information) => Task.FromResult(true);

        public async override Task<Data?> Process(Information information)
        {

#if DEBUG

            // Parse the input for the template id and user data

            int firstSpace = information.Input?.Raw?.IndexOf(' ') ?? -1;

            if (firstSpace > 6)
            {
                var templateId = information.Input?.Raw?.Substring(6, firstSpace - 6);
                var userData = information.Input?.Raw?.Substring(firstSpace + 1);

                if (string.IsNullOrEmpty(templateId) || !_agent.Catalog.ContainsKey(templateId) || string.IsNullOrEmpty(userData))
                {
                    return null;
                }

                Template template = _agent.Catalog[templateId];

                Data data;

                if (template.InputKeys != null && template.InputKeys.Length > 0)
                {
                    data = new Data(userData, DataFormat.STRUCTURED);
                }
                else
                {
                    data = new Data(userData, DataFormat.RAW);
                }

                return await information.Publish(templateId, data);
            }
            else
            {
                // TODO: Allow parameterless debug with no data
                return new Data("Not Supported");
            }

#endif
            return new Data("Debug not enabled");
        }
    }
}

File: Agents/Interaction/python_templates/debug.py
from technologai_agent.information import Information
from technologai_agent.template import Template

import json

class Debug(Template):
    def __init__(self, agent, **kwargs):
        super().__init__(id="debug", description="Debug", **kwargs)
        self._agent = agent

    async def assess(self, information: Information):
        return True

    # TODO in the C# code, there is a IF debug, but I am not sure if Python has an equivalent
    async def process(self, information: Information):
        # Parse the input for the template id and user data
        first_space = information.input.find(' ')

        if first_space > 6:
            template_id = information.input[6:first_space]
            user_data = information.input[first_space + 1:]

            if not template_id or template_id not in self._agent.catalog or not user_data:
                return None

            template = self._agent.catalog[template_id]

            # if template.input_keys and len(template.input_keys) > 0:
            #     data = Data(user_data, DataFormat.STRUCTURED)
            # else:
            #     data = Data(user_data, DataFormat.RAW)
            if template.input_keys and len(template.input_keys) == 0:
                user_data = json.loads(user_data)

            return await information.publish(template_id, user_data)
        else:
            # TODO: Allow parameterless debug with no data
            return "Invalid debug command"
